app/tests/reader_studies_tests/test_api.py

```diff
@@ -294,3 +298,157 @@ def test_answer_is_correct_type(client, answer_type, answer, expected):
+    )
+
+    response = get_view_for_user(
+        viewname="api:reader-study-export-questions",
+        reverse_kwargs={"pk": rs.pk},
+        user=editor,
+        client=client,
+        method=client.get,
+        content_type="application/json",
+    )
+
+    headers = str(response.serialize_headers())
+    content = str(response.content)
+
+    assert response.status_code == 200
+    assert "Content-Type: text/csv" in headers
+    assert f'filename="{rs.slug}-questions.csv"' in headers
+    assert q.question_text in content
+    assert q.get_answer_type_display() in content
+
+    response = get_view_for_user(
+        viewname="api:reader-study-export-questions",
+        reverse_kwargs={"pk": rs.pk},
+        user=reader,
+        client=client,
+        method=client.get,
+        content_type="application/json",
+    )
+    assert response.status_code == 404
+
+    a = AnswerFactory(question=q, answer=answer)
+
+    response = get_view_for_user(
+        viewname="api:reader-study-export-answers",
+        reverse_kwargs={"pk": rs.pk},
+        user=editor,
+        client=client,
+        method=client.get,
+        content_type="application/json",
+    )
+
+    headers = str(response.serialize_headers())
+    content = str(response.content)
+
+    assert response.status_code == 200
+    assert "Content-Type: text/csv" in headers
+    assert f'filename="{rs.slug}-answers.csv"' in headers
+    assert a.creator.username in content
+    assert a.question.question_text in content
+    if isinstance(answer, dict):
+        for key in answer:
+            assert key in content
+    else:
+        assert re.sub(r"[\n\r\t]", " ", str(a.answer)) in content
+
+    response = get_view_for_user(
+        viewname="api:reader-study-export-answers",
+        reverse_kwargs={"pk": rs.pk},
+        user=reader,
+        client=client,
+        method=client.get,
+        content_type="application/json",
+    )
+    assert response.status_code == 404
+
+
+@pytest.mark.parametrize(
+    "data,elements,lines",
+    (
+        ([["a"], ["b"], ["c"]], 1, 3),
+        ([["a"], ["b,c"], ["c"]], 1, 3),
+        ([["a\nb"], ["b"], ["c"]], 1, 3),
+        ([["a\rb\nc", "\nb", "\rc\r\r"]], 3, 1),
+        ([["a", "a", "\na"], ["b", "b", "b"], ["c", "c", "c"]], 3, 3),
+        (
+            [["a", '{"a":\n{"b": "c\nd"}\n}'], ["b", "b,c,d"], ["c", "d\r"]],
+            2,
+            3,
+        ),
+    ),
+)
+def test_csv_export_preprocessing(data, elements, lines):
+    exporter = ExportCSVMixin()
+    processed = exporter._preprocess_data(data)
+    assert len(processed) == lines
+
+    # Unfortunately, we have to create an actual file here, as both tempfile
+    # and StringIO seem to cause issues with line endings
+    with open("/tmp/csv.csv", "w+", newline="") as f:
+        writer = csv.writer(f)
+        writer.writerows(processed)
+
+    with open("/tmp/csv.csv", "r", newline="") as f:
```

Comment:
`/tmp/` should not be used directly - better to use the `tmp_path` fixture from pytest http://doc.pytest.org/en/latest/tmpdir.html#the-tmp-path-fixture

Reply:
Changed